client/src/plus/collections/frequently-viewed.tsx

```diff
@@ -0,0 +1,208 @@
+import { useEffect, useState } from "react";
+import { Doc, DocParent } from "../../../../libs/types/document";
+import { Collection } from "./api";
+
+export interface FrequentlyViewedCollection extends Collection {
+  items: FrequentlyViewedEntry[];
+}
+
+export type FrequentlyViewedEntry = {
+  parents?: DocParent[];
+  serial: number;
+  summary?: string;
+  timestamp?: number;
+  timestamps: number[];
+  title: string;
+  url: string;
+  visitCount: number;
+};
+
+const FREQUENTLY_VIEWED_STORAGE_KEY = "frequently-viewed-documents";
+
+const ThirtyDaysMilliseconds = 30 * 24 * 60 * 60 * 1000;
+const isWithinLastThirtyDays = (date: Date): boolean => {
+  const currentDate = new Date().getTime();
+  return date.getTime() >= currentDate - ThirtyDaysMilliseconds;
+};
+
+function getFrequentlyViewed(): FrequentlyViewedEntry[] {
+  let frequentlyViewed: string | null = null;
+  try {
+    frequentlyViewed = localStorage.getItem(FREQUENTLY_VIEWED_STORAGE_KEY);
+  } catch (err) {
+    console.warn(
+      "Unable to read frequently viewed documents from localStorage",
+      err
+    );
+  }
+
+  const entries = JSON.parse(
+    frequentlyViewed || "[]"
+  ) as FrequentlyViewedEntry[];
+
+  // Migrate old entries just once. Log their most recent timestamp 'visitCount' times. (If within last 30 days)
+  entries.forEach((val) => {
+    if (!val.timestamps) {
+      val.timestamps = [];
+      if (val.timestamp) {
+        if (isWithinLastThirtyDays(new Date(val.timestamp))) {
+          for (let i = 0; i < val.visitCount; i++) {
+            val.timestamps.push(val.timestamp);
+          }
+          //Delete old timestamp
+          val.timestamp = undefined;
+        }
+      }
+    }
+  });
+
+  // Assign serials to old entries.
+  entries.forEach((e) => {
+    e.serial =
+      e.serial === undefined
+        ? getNextFrequentlyViewedSerial(entries)
+        : e.serial;
+  });
+  //Remove all timestamps older than 30 days and any pages with no more hits.
+  return filterFrequentlyViewed(entries);
+}
+
+function setFrequentlyViewed(
+  frequentlyViewed: FrequentlyViewedEntry[],
+  done = () => {}
+) {
+  try {
+    localStorage.setItem(
+      FREQUENTLY_VIEWED_STORAGE_KEY,
+      JSON.stringify(frequentlyViewed)
+    );
+    done();
+  } catch (err) {
+    console.warn(
+      "Failed to write frequently viewed documents to localStorage",
+      err
+    );
+  }
+}
+
+const sortByTimestampThenVistsDesc = (
```

Comment:
Shouldn't this be the other way around? Sort by visits then timestamp? It's highly unlikely we'll ever get two timestamps to be the same - indeed, if they were, I think we'd hit a race condition and only one would save. And regardless - this is frequently viewed (implies most frequently viewed being at the top) not recently viewed.

Reply:
The timestamps are inserted in order of most recent first. I expected most recent to trump many visits. The alternative form we could end up with a fairly static looking top 10 for 30 days.  Let me check with our product folks